src/components/Toolbar.js

```diff
@@ -11,13 +37,87 @@ const shortcutIds = [
   'search',
 ]
 
-export const Toolbar = connect(({ settings: { dark } }) => ({ dark }))(({ dark }) =>
-  <div className="toolbar">
-    {shortcutIds.map(id => {
-      const { name, svg: Icon, exec } = shortcutById(id)
-      return <div key={id} className="toolbar-icon" title={name} onClick={exec}>
-        <Icon fill={dark ? 'white' : 'black'} />
+const ToolbarMessageOverlay = ({
+  classname = 'toolbar-overlay',
+  name,
+  description
+}) => {
+  return (
+    <div className={classname}>
+      <div className={'overlay-name'}>{name}</div>
+      <div className={'overlay-body'}>{description}</div>
+    </div>
+  )
+}
+
+let holdTimer
+
+const onHoldDownShortcut = id => {
+  // on chrome setTimeout doesn't seem to work on the first click, clearing it before hand fixes the problem
+  clearTimeout(holdTimer)
+  holdTimer = setTimeout(() => {
+    overlayReveal(id)
+  }, 500)
+}
+
+const clearHoldTimer = () => {
+  overlayHide()
+  clearTimeout(holdTimer)
+}
+
+export const Toolbar = connect(({ toolbarOverlay, settings: { dark } }) => ({
+  dark,
+  toolbarOverlay
+}))(
+  ({
+    dark,
+    toolbarOverlay: {
+      showOverlay,
+      shortcutId
+    }
+  }) => (
+    <div>
+      <div className={'toolbar-container'}>
+        <div className={'toolbar'}>
+          {shortcutIds.map(id => {
+            const { name, svg: Icon, exec } = shortcutById(id)
+            return (
+              <div
+                key={name}
+                id={id}
+                className='toolbar-icon'
+                onMouseDown={() => onHoldDownShortcut(id)}
+                onMouseOver={() => overlayUpdate(id)}
+                onTouchStart={() => onHoldDownShortcut(id)}
+                onClick={() => exec(id)}
+              >
+                <Icon id={id} fill={dark ? 'white' : 'black'} />
+              </div>
+            )
+          })}
+        </div>
+        <TransitionGroup>
+          <CSSTransition key={0} timeout={200} classNames='fade'>
+            {!isTouchEnabled() && showOverlay ? (
+              <ToolbarMessageOverlay
+                name={shortcutById(shortcutId).name}
+                description={shortcutById(shortcutId).description}
```

Comment:
It would be more efficient to call `shortcutById` once to retrieve the shortcut object, then access `id`, `name`, and `description` from the object while it is in scope.

Reply:
Got it.